{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f39df9e-7a2e-4524-928c-691358aaf2f7",
   "metadata": {},
   "source": [
    "# BIO 265 Final Project - Lily Kalcec, Pre Lavania, and Nicole Haseley"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f0506ef9-9df5-4940-8dc6-7714a3085b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 Import libraries\n",
    "# standard libraries\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import scipy\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "# differential expression\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "# dimentionality reduction and clustering\n",
    "from sklearn.decomposition import PCA\n",
    "import sklearn.cluster as cluster\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import umap\n",
    "\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "21f0787b-6aa4-400b-8488-7edeaa2c0e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2 Load in data (final RNAseq readcounts)\n",
    "RNAseq_data = pd.read_csv(\"rawcounts-final.csv\")\n",
    "RNAseq_data.drop(\"ensemble_id_version\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c612ea57-614d-4c05-bb60-0d0e9521e291",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CFB2001</th>\n",
       "      <th>CFB2006</th>\n",
       "      <th>CFB2007</th>\n",
       "      <th>CFB2009</th>\n",
       "      <th>CFB2013</th>\n",
       "      <th>CFB2022</th>\n",
       "      <th>CFB2039</th>\n",
       "      <th>CFB2044</th>\n",
       "      <th>CFB2045</th>\n",
       "      <th>CFB2049</th>\n",
       "      <th>...</th>\n",
       "      <th>CFB2146</th>\n",
       "      <th>CFB2149</th>\n",
       "      <th>CFB2152</th>\n",
       "      <th>CFB2159</th>\n",
       "      <th>CFB2160</th>\n",
       "      <th>CFB2163</th>\n",
       "      <th>CFB2166</th>\n",
       "      <th>CFB2169</th>\n",
       "      <th>CFB2171</th>\n",
       "      <th>CFB2186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>741</td>\n",
       "      <td>260</td>\n",
       "      <td>681</td>\n",
       "      <td>1110</td>\n",
       "      <td>640</td>\n",
       "      <td>754</td>\n",
       "      <td>965</td>\n",
       "      <td>1107</td>\n",
       "      <td>906</td>\n",
       "      <td>1105</td>\n",
       "      <td>...</td>\n",
       "      <td>1000</td>\n",
       "      <td>680</td>\n",
       "      <td>533</td>\n",
       "      <td>799</td>\n",
       "      <td>1441</td>\n",
       "      <td>705</td>\n",
       "      <td>1000</td>\n",
       "      <td>1452</td>\n",
       "      <td>1140</td>\n",
       "      <td>1875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>881</td>\n",
       "      <td>293</td>\n",
       "      <td>782</td>\n",
       "      <td>1013</td>\n",
       "      <td>581</td>\n",
       "      <td>587</td>\n",
       "      <td>884</td>\n",
       "      <td>770</td>\n",
       "      <td>711</td>\n",
       "      <td>1049</td>\n",
       "      <td>...</td>\n",
       "      <td>882</td>\n",
       "      <td>603</td>\n",
       "      <td>512</td>\n",
       "      <td>560</td>\n",
       "      <td>1240</td>\n",
       "      <td>619</td>\n",
       "      <td>763</td>\n",
       "      <td>1504</td>\n",
       "      <td>1011</td>\n",
       "      <td>1421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>30</td>\n",
       "      <td>71</td>\n",
       "      <td>102</td>\n",
       "      <td>72</td>\n",
       "      <td>89</td>\n",
       "      <td>121</td>\n",
       "      <td>124</td>\n",
       "      <td>83</td>\n",
       "      <td>122</td>\n",
       "      <td>...</td>\n",
       "      <td>122</td>\n",
       "      <td>76</td>\n",
       "      <td>40</td>\n",
       "      <td>78</td>\n",
       "      <td>157</td>\n",
       "      <td>85</td>\n",
       "      <td>116</td>\n",
       "      <td>130</td>\n",
       "      <td>117</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60644</th>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>38</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>23</td>\n",
       "      <td>34</td>\n",
       "      <td>26</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>45</td>\n",
       "      <td>35</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60645</th>\n",
       "      <td>2511</td>\n",
       "      <td>1032</td>\n",
       "      <td>2328</td>\n",
       "      <td>2841</td>\n",
       "      <td>3197</td>\n",
       "      <td>1091</td>\n",
       "      <td>3035</td>\n",
       "      <td>1812</td>\n",
       "      <td>1845</td>\n",
       "      <td>3294</td>\n",
       "      <td>...</td>\n",
       "      <td>1773</td>\n",
       "      <td>1722</td>\n",
       "      <td>728</td>\n",
       "      <td>1649</td>\n",
       "      <td>3436</td>\n",
       "      <td>2090</td>\n",
       "      <td>1497</td>\n",
       "      <td>4886</td>\n",
       "      <td>2635</td>\n",
       "      <td>2789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60646</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60647</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60648</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60649 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CFB2001  CFB2006  CFB2007  CFB2009  CFB2013  CFB2022  CFB2039  CFB2044  \\\n",
       "0            1        4       30       15        4        7       23       19   \n",
       "1            0        0        0        0        0        0        0        0   \n",
       "2          741      260      681     1110      640      754      965     1107   \n",
       "3          881      293      782     1013      581      587      884      770   \n",
       "4           64       30       71      102       72       89      121      124   \n",
       "...        ...      ...      ...      ...      ...      ...      ...      ...   \n",
       "60644       14       13       25       38       12       21       23       34   \n",
       "60645     2511     1032     2328     2841     3197     1091     3035     1812   \n",
       "60646        0        0        0        0        0        0        0        0   \n",
       "60647        0        0        0        0        0        0        0        0   \n",
       "60648        0        0        0        0        0        0        0        0   \n",
       "\n",
       "       CFB2045  CFB2049  ...  CFB2146  CFB2149  CFB2152  CFB2159  CFB2160  \\\n",
       "0           15       15  ...       24       12       14       12       16   \n",
       "1            0        0  ...        0        0        0        0        0   \n",
       "2          906     1105  ...     1000      680      533      799     1441   \n",
       "3          711     1049  ...      882      603      512      560     1240   \n",
       "4           83      122  ...      122       76       40       78      157   \n",
       "...        ...      ...  ...      ...      ...      ...      ...      ...   \n",
       "60644       26       36  ...       27       20       12       10       28   \n",
       "60645     1845     3294  ...     1773     1722      728     1649     3436   \n",
       "60646        0        0  ...        0        0        0        0        0   \n",
       "60647        0        0  ...        1        0        0        0        0   \n",
       "60648        0        0  ...        0        0        0        0        0   \n",
       "\n",
       "       CFB2163  CFB2166  CFB2169  CFB2171  CFB2186  \n",
       "0            7       20       23       20       23  \n",
       "1            0        0        0        0        0  \n",
       "2          705     1000     1452     1140     1875  \n",
       "3          619      763     1504     1011     1421  \n",
       "4           85      116      130      117      157  \n",
       "...        ...      ...      ...      ...      ...  \n",
       "60644       27       27       45       35       44  \n",
       "60645     2090     1497     4886     2635     2789  \n",
       "60646        0        0        0        0        0  \n",
       "60647        0        0        0        0        0  \n",
       "60648        0        0        0        0        0  \n",
       "\n",
       "[60649 rows x 42 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNAseq_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f817fe4-cb38-4753-8c6c-49fff6a225e3",
   "metadata": {},
   "source": [
    "# Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "494b42d5-62f2-464d-8e11-ba9c9d6e19aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3 separate the samples into control and test (have ntm disease) groups\n",
    "control_samples = [\"CFB2006\", \"CFB2007\", \"CFB2009\", \"CFB2013\", \"CFB2022\", \"CFB2039\", \"CFB2044\", \"CFB2058\", \n",
    "                  \"CFB2059\", \"CFB2074\", \"CFB2075\", \"CFB2082\", \"CFB2087\", \"CFB2088\", \"CFB2102\", \"CFB2108\", \n",
    "                  \"CFB2112\", \"CFB2127\", \"CFB2134\", \"CFB2138\", \"CFB2140\", \"CFB2142\", \"CFB2146\", \"CFB2149\", \n",
    "                   \"CFB2152\", \"CFB2160\", \"CFB2166\", \"CFB2169\", \"CFB2171\", \"CFB2186\"] \n",
    "ntm_samples = [\"CFB2001\", \"CFB2045\", \"CFB2049\", \"CFB2060\", \"CFB2061\", \"CFB2071\", \"CFB2084\", \"CFB2098\",\n",
    "              \"CFB2105\", \"CFB2128\", \"CFB2159\", \"CFB2163\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "71a7abf1-68b7-4954-803e-d3345f2a24f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4\n",
    "# create a dictionary mapping each sample to its condition\n",
    "metadata_dict = {sample: \"Control\" for sample in control_samples}\n",
    "metadata_dict.update({sample: \"NTM\" for sample in ntm_samples})\n",
    "\n",
    "# convert to dataframe\n",
    "metadata = pd.DataFrame.from_dict(metadata_dict, orient=\"index\", columns=[\"condition\"])\n",
    "\n",
    "# ensure the index matches the column names of count_data\n",
    "metadata.index.name = \"sample\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "47edc753-d7c6-42de-8985-6f09b2cea6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5\n",
    "metadata = metadata.loc[RNAseq_data.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f18baaf7-83a0-48dd-9ed6-16813d02cb91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6\n",
    "metadata['condition'] = metadata['condition'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0ac0a2-ec6a-47d7-bb7d-75161867c1bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.12/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/var/folders/km/jx84sy451dv01dcs0m4gcrp80000gn/T/ipykernel_2367/2599395339.py:3: DeprecationWarning: ref_level is deprecated and no longer has any effect. It will beremoved in a future release.\n",
      "  dds = DeseqDataSet(\n",
      "/var/folders/km/jx84sy451dv01dcs0m4gcrp80000gn/T/ipykernel_2367/2599395339.py:3: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "  dds = DeseqDataSet(\n",
      "Fitting size factors...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.41 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 26.00 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 1.23 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 20.47 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 13.74 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.27 seconds.\n",
      "\n",
      "Replacing 134 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.21 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.22 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.20 seconds.\n",
      "\n",
      "Running Wald tests...\n"
     ]
    }
   ],
   "source": [
    "# 7\n",
    "# initialize PyDESeq2 model\n",
    "dds = DeseqDataSet(\n",
    "    counts=RNAseq_data.T,  # PyDESeq2 expects samples as rows, genes as columns\n",
    "    metadata=metadata,\n",
    "    design_factors=[\"condition\"],  # Needs to be a list\n",
    "    ref_level=\"Control\"  # Setting \"Control\" as the reference group\n",
    ")\n",
    "\n",
    "# run DESeq2 analysis\n",
    "dds.deseq2()\n",
    "\n",
    "# extract results\n",
    "res = DeseqStats(dds, contrast=[\"condition\", \"Control\", \"NTM\"], alpha=0.05)\n",
    "res.summary()\n",
    "\n",
    "# get the results as a DataFrame\n",
    "DE_results_df = res.results_df\n",
    "display(DE_results_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7a3939-9461-424e-aeda-56a6a7fc52f7",
   "metadata": {},
   "source": [
    "# Raw p-values – applying an arbitrary **p-value cutoff of 0.001 and fold change cutoff of 0.5** for signficantly differentially expressed genes (as authors did)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c294120b-191a-4161-9937-d342b88bbe4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8\n",
    "#### Using Raw p-values and fold change from DE_results_df (data frame with results from the PyDESeq2 analysis)\n",
    "\n",
    "# Set p-value threshold and log2 fold-change threshold\n",
    "pvalue_cutoff = 0.001\n",
    "log2fc_cutoff = 0.5  # Absolute log2 fold-change threshold\n",
    "\n",
    "# Remove NaN p-values and create a new DataFrame copy\n",
    "DE_results_df = DE_results_df.dropna(subset=['pvalue']).copy()\n",
    "\n",
    "# Compute -log10(p-values)\n",
    "DE_results_df.loc[:, 'log_pvalue'] = -np.log10(DE_results_df[\"pvalue\"])\n",
    "\n",
    "# Flag significant genes based on BOTH p-value and log2 fold-change threshold\n",
    "DE_results_df.loc[:, 'significant'] = (DE_results_df[\"pvalue\"] <= pvalue_cutoff) & (DE_results_df[\"log2FoldChange\"].abs() >= log2fc_cutoff)\n",
    "\n",
    "# Identify upregulated and downregulated genes\n",
    "DE_results_df.loc[:, 'upregulated'] = DE_results_df['significant'] & (DE_results_df['log2FoldChange'] > log2fc_cutoff)\n",
    "DE_results_df.loc[:, 'downregulated'] = DE_results_df['significant'] & (DE_results_df['log2FoldChange'] < -log2fc_cutoff)\n",
    "\n",
    "# Convert p-values to -log10 scale for plotting\n",
    "neg_log_pvals = DE_results_df['log_pvalue']\n",
    "\n",
    "# Create volcano plot\n",
    "plt.figure(figsize=(6, 5))\n",
    "\n",
    "# Plot non-significant genes in gray\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][~DE_results_df[\"significant\"]],\n",
    "    neg_log_pvals[~DE_results_df[\"significant\"]],\n",
    "    c='gray', alpha=0.5, s=10, label=\"Not Significant\"\n",
    ")\n",
    "\n",
    "# Plot upregulated genes in red\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][DE_results_df[\"upregulated\"]],\n",
    "    neg_log_pvals[DE_results_df[\"upregulated\"]],\n",
    "    c='red', alpha=0.7, s=15, label=\"Upregulated\"\n",
    ")\n",
    "\n",
    "# Plot downregulated genes in green\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][DE_results_df[\"downregulated\"]],\n",
    "    neg_log_pvals[DE_results_df[\"downregulated\"]],\n",
    "    c='limegreen', alpha=0.7, s=15, label=\"Downregulated\"\n",
    ")\n",
    "\n",
    "# Add threshold lines\n",
    "plt.axhline(y=-np.log10(pvalue_cutoff), color='red', linestyle='--', linewidth=1)\n",
    "plt.axvline(x=log2fc_cutoff, color='red', linestyle='--', linewidth=1)\n",
    "plt.axvline(x=-log2fc_cutoff, color='red', linestyle='--', linewidth=1)\n",
    "\n",
    "# Labels and title\n",
    "plt.xlabel(\"log2FoldChange\")\n",
    "plt.ylabel(\"-log10(p-value)\")\n",
    "plt.title(\"Volcano Plot with p-value & Fold-Change Cutoff\")\n",
    "plt.legend(fontsize=9)\n",
    "plt.show()\n",
    "\n",
    "# Print the number of significant genes\n",
    "significant_genes = DE_results_df[DE_results_df['significant']]\n",
    "print(f\"Number of significant genes (p-value ≤ {pvalue_cutoff}, |log2FC| ≥ {log2fc_cutoff}): {len(significant_genes)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "129b83a1-e8e4-43b2-990f-dc32df300e35",
   "metadata": {},
   "source": [
    "# Benjamini-Hochberg corrrected p-values – also applying an arbitrary **p-value cutoff of 0.001 and fold change cutoff of 0.5** for signficantly differentially expressed genes (as authors did)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ff8714a-7bc4-47fa-bff3-33e02d67bd20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9\n",
    "#### With Benjamini-Hochberg (BH) correction - using BH-corrected p-values (\"padj\" values) from the same DE_results_df analysis\n",
    "\n",
    "# Set p-value threshold and log2 fold-change threshold\n",
    "padj_threshold = 0.001\n",
    "log2fc_threshold = 0.5 # Absolute log2 fold-change threshold\n",
    "\n",
    "# Remove NaN padj values\n",
    "DE_results_df = DE_results_df.dropna(subset=['padj'])\n",
    "\n",
    "# Compute -log10(padj)\n",
    "DE_results_df['log_padj'] = -np.log10(DE_results_df['padj'])\n",
    "\n",
    "# Determine significance: Significant if padj < threshold and |log2FoldChange| > threshold\n",
    "DE_results_df['significant'] = (DE_results_df['padj'] < padj_threshold) & (np.abs(DE_results_df['log2FoldChange']) > log2fc_threshold)\n",
    "\n",
    "# Create volcano plot\n",
    "plt.figure(figsize=(6, 5))\n",
    "\n",
    "# Plot non-significant genes in gray\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][~DE_results_df[\"significant\"]],\n",
    "    DE_results_df['log_padj'][~DE_results_df[\"significant\"]],\n",
    "    c='gray', alpha=0.5, s=10, label=\"Not Significant\")\n",
    "\n",
    "# Plot significant upregulated genes in red\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][(DE_results_df[\"significant\"]) & (DE_results_df[\"log2FoldChange\"] > 0)],\n",
    "    DE_results_df['log_padj'][(DE_results_df[\"significant\"]) & (DE_results_df[\"log2FoldChange\"] > 0)],\n",
    "    c='red', alpha=0.7, s=15, label=\"Upregulated\")\n",
    "\n",
    "# Plot significant downregulated genes in green\n",
    "plt.scatter(\n",
    "    DE_results_df[\"log2FoldChange\"][(DE_results_df[\"significant\"]) & (DE_results_df[\"log2FoldChange\"] < 0)],\n",
    "    DE_results_df['log_padj'][(DE_results_df[\"significant\"]) & (DE_results_df[\"log2FoldChange\"] < 0)],\n",
    "    c='green', alpha=0.7, s=15, label=\"Downregulated\")\n",
    "\n",
    "# Labels and title\n",
    "plt.xlabel(\"log2FoldChange\")\n",
    "plt.ylabel(\"-log10(padj)\")\n",
    "plt.title(\"Volcano Plot (padj < 0.001, |log2FC| > 0.5)\")\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
